docs/design/rubygems.md

```diff
@@ -0,0 +1,357 @@
+That's why we **do not use bundler** to download dependencies.
+Instead, as stated above, we parse the `Gemfile.lock` file directly and download the gems from <https://rubygems.org>.
+
+**Note**: parsing `Gemfile.lock` is done via [gemlock-parser](https://github.com/containerbuildsystem/gemlock-parser),
+which is vendored from [scancode-toolkit](https://github.com/nexB/scancode-toolkit/blob/develop/src/packagedcode/gemfile_lock.py).
+
+`Gemfile` example:
+
+```ruby
+source 'https://rubygems.org'
+
+gem 'rails', '= 6.1.7'
+
+system('echo "Hello, world!"')
+system('sudo rm -rf /')
+```
+
+Source code for bundler lockfile parsing:
+
+<https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/lockfile_parser.rb>
+
+### Missing features
+
+Bundler is not pinned as a dependency with version in the `Gemfile.lock` (even if it is pinned in the `Gemfile`).
+It only appears in the `BUNDLED WITH` section in the `Gemfile.lock` file.
+But it is important for the same version of bundler to be installable and used for resolving dependencies.
+Using the bundler from the build image usually does not fit.
+
+## Dependencies
+
+There four types of [sources](https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/lockfile_parser.rb#L48) for dependencies in the `Gemfile.lock` file:
+
+### Gem dependencies
+
+Regular gem dependencies located at source url, ~~usually~~ in our case always <https://rubygems.org>.
+Each gem can be accessed by its name and version - rubygems.org/gems/`<name>`-`<version>`.gem
+
+Example of a gem dependency in the `Gemfile.lock` file:
+
+```Gemfile.lock
+...
+GEM
+  remote: https://rubygems.org/
+  specs:
+    ...
+    rails (6.1.4)
+      # transitive dependencies
+      actioncable (= 6.1.4)
+      actionmailbox (= 6.1.4)
+      actionmailer (= 6.1.4)
+      actionpack (= 6.1.4)
+      actiontext (= 6.1.4)
+      actionview (= 6.1.4)
+      activejob (= 6.1.4)
+      activemodel (= 6.1.4)
+      activerecord (= 6.1.4)
+      activestorage (= 6.1.4)
+      activesupport (= 6.1.4)
+      bundler (>= 1.15.0)
+      railties (= 6.1.4)
+      sprockets-rails (>= 2.0.0)
+...
+```
+
+### Git dependencies
+
+Example of a git dependency in the `Gemfile.lock` file:
+
+```Gemfile.lock
+...
+GIT
+  remote: https://github.com/porta.git
+  revision: 779beabd653afcd03c4468e0a69dc043f3bbb748
+  branch: main
+  specs:
+    porta (2.14.1)
+...
+```
+
+Bundler uses this format to cache git repositories:
+
+[https://github.com/rubygems/.../bundler/source/git.rb#L139](https://github.com/rubygems/rubygems/blob/3da9b1dda0824d1d770780352bb1d3f287cb2df5/bundler/lib/bundler/source/git.rb#L139>)
+
+Any other format will cause bundler to re-download the repository -> cache invalidation -> the build will fail.
+
+### Path dependencies
+
+Example of a path dependency in the `Gemfile.lock` file:
+
+```Gemfile.lock
+...
+PATH
+  remote: ../foobar
+  specs:
+    foobar (0.1.0)
+...
+```
+
+### Plugins
+
+TODO
```

Comment:
IMO we just need to call out that `Plugin`-type dependencies are not supported. 

I _would_ be curious to know what they are though :grin: